### ISSUE TITLE
Can't build on UNIX

### ISSUE THREAD

**Issue by JunielKatarn** [Microsoft/msbuild#567]
Hi. I can't build the latest xplat commit (See 5a2dcc67bff636fa8107bbf08e1852a2e34121b8) on Mac OS X 10.11.

After checking out the latest xplat, I run `./cibuild` and always get the following output:
```
Unsupported target detected: . Configuring as if for CoreCLR
Unsupported host detected: . Configuring as if for CoreCLR
Installing dotnet cli...
Restoring BuildTools version 1.0.25-prerelease-00231-07...
ERROR: Could not restore build tools correctly. See '/Volumes/DATA/Build/msb/buildme/init-tools.log' for more details.
Initializing BuildTools...
./init-tools.sh: line 84: /Volumes/DATA/Build/msb/buildme/packages/Microsoft.DotNet.BuildTools/1.0.25-prerelease-00231-07/lib/init-tools.sh: No such file or directory
Done initializing tools.
Build Command: ./Tools/corerun  ./Tools/MSBuild.exe ./build.proj /t:Build /p:OS=OSX /p:Configuration=Debug-NetCore /verbosity:minimal  /m /fl  "/flp:v=diag;logfile=./msbuild.log"
./cibuild.sh: line 42: ./Tools/corerun: No such file or directory
```

After looking at `init-tools.log`, found the following:
```
unning: /Volumes/DATA/Build/msb/buildme/init-tools.sh
Installing 'https://dotnetcli.blob.core.windows.net/dotnet/beta/Binaries/1.0.0-beta-002173/dotnet-dev-osx-x64.1.0.0-beta-002173.tar.gz' to '/Volumes/DATA/Build/msb/buildme/Tools/dotnetcli/dotnet.tar'
Running: /Volumes/DATA/Build/msb/buildme/Tools/dotnetcli/dotnet restore "/Volumes/DATA/Build/msb/buildme/Tools/1.0.25-prerelease-00231-07/project.json" --packages /Volumes/DATA/Build/msb/buildme/packages --source https://dotnet.myget.org/F/dotnet-buildtools/api/v3/index.json
log  : Restoring packages for /Volumes/DATA/Build/msb/buildme/Tools/1.0.25-prerelease-00231-07/project.json...
error: Failed to retrieve information from remote source '/Volumes/DATA/Build/msb/buildme/packages'.
error: Unable to load the service index for source https://dotnet.myget.org/F/dotnet-buildtools/api/v3/index.json.
error:   The type initializer for 'Crypto' threw an exception.
error:   Unable to load DLL 'System.Security.Cryptography.Native': The specified module could not be found.
error:    (Exception from HRESULT: 0x8007007E)
Running: /Volumes/DATA/Build/msb/buildme/packages/Microsoft.DotNet.BuildTools/1.0.25-prerelease-00231-07/lib/init-tools.sh /Volumes/DATA/Build/msb/buildme /Volumes/DATA/Build/msb/buildme/Tools/dotnetcli/dotnet /Volumes/DATA/Build/msb/buildme/Tools
```

I notice a few things:
1. The `packages` subdirectory never got created.
2. I can browse to `https://dotnet.myget.org/F/dotnet-buildtools/api/v3/index.json` even when the scripts says its index can't be loaded.
3. Maybe package `System.Security.Cryptography.Native` is outdated in the dependency mappings?
4. The same commit was built successfully on Jenkins at http://dotnet-ci.cloudapp.net/job/Microsoft_msbuild/job/innerloop_xplat_OSX/87/ .

Is there any pre-configuration I need to make in order to build locally?

**Comment by JunielKatarn** [comment]
I just verified the issue also appears on Ubuntu.
I ran `./cibuild.sh --scope Compile` (same as latest successful build at http://dotnet-ci.cloudapp.net/job/Microsoft_msbuild/job/innerloop_xplat_Ubuntu/lastBuild/console), and found the same error.

CoreRun is not being downloaded/installed.

**Comment by rainersigwald** [comment]
That's interesting; like the CI servers I can build on both OS X and Linux. Sadly, `ERROR_MOD_NOT_FOUND` is not particularly informative!

I don't think this is related to our `project.json`, since it appears to be a crypto failure when connecting over HTTP. It looks to me like you have a bad on-disk deployment of the `dotnet` CLI. But I don't know why that would be failing for you.

Do you have a `System.Security.Cryptography.Native.so` anywhere under `Tools/dotnetcli` (on Ubuntu--on OS X it's `.dylib`)?

**Comment by radical** [comment]
On a fresh install of `Ubuntu 15.10`, it seems to failing running `dotnet` with ..

`Failed to initialize CoreCLR, HRESULT: 0x80131500`

Looking at it with strace, it seems to be looking for `libicuuc.so.52` and `libicui18n.so.52`, and at least on `Ubuntu 15.10` that I tried this on, `libicu*so.55` is available.

But I don't have any trouble building on OSX 10.11.4 , but my OSX install might not really be "clean"!

**Comment by akoeplinger** [comment]
@radical Ubuntu 15.10 isn't supported by .NET Core/CLI yet (https://github.com/dotnet/cli/issues/2072), so this is expected.

**Comment by radical** [comment]
@JunielKatarn which version of Ubuntu are you running?

**Comment by radical** [comment]
For OSX, I would `rm -Rf bin/ Tools/ packages/` and then try the build again.

**Comment by JunielKatarn** [comment]
@radical It's 15.10. I guess it's not supported, then.
Also, for OS X (10.11.4), I tried using a clean checkout (nothing built previously).
Fails every time.

Maybe I misread instructions? Do I need to install something before?
Or, should the `cibuild.sh` script resolve/fetch all dependencies by itself?

**Comment by rainersigwald** [comment]
My Mac is also not clean, but I'm pretty sure `cibuild.sh` should do everything you need. Do you see `System.Security.Cryptography.Native.dylib` under the `Tools/dotnetcli` folder when starting from scratch on OS X?

**Comment by JunielKatarn** [comment]
I do see that dylib under `./Tools/dotnetcli/shared/Microsoft.NETCore.App/1.0.0-rc2-23931/System.Security.Cryptography.Native.dylib`.

I'm afraid cibuild.sh is not working for me :(
If it's any use, I here is a ZIP file containing the generated `Tools` subdirectory, along with the build log:
https://www.dropbox.com/sh/var5r5pbn5hwn38/AAAKitK_BwuWKvF9121TlmLma

**Comment by Sarabeth-Jaffe-Microsoft** [comment]
@JunielKatarn Any progress resolving this issue? I'm running into it myself.

**Comment by JunielKatarn** [comment]
No progress, I'm afraid. I run into this issue both in OS X (El Capitan), and the last 3 releases of Ubuntu (including brand new 16.04).

In short, the .NET tools are not getting downloaded correctly.

@cdmihai, @rainersigwald, do you guys have any pointers?

**Comment by radical** [comment]
Try to update to the xplat HEAD, and try again. Your logs suggest that it is trying to fetch an older version:

`EXEC : warning : Unable to find version '1.2.0-beta-20151112-01' of package 'Microsoft.Net.Compilers'`

This was updated in https://github.com/Microsoft/msbuild/commit/9dd9195913e49459d40e9e5c38977f52a856a09a

**Comment by radical** [comment]
Your original post says that you are building - 5a2dcc6 but still getting the older version? I would update to the latest xplat, `git-clean -xdf` my checkout and try again.

**Comment by JunielKatarn** [comment]
@radical Updating to the head commit and cleaning the local repository goes without saying :). It's always my first course of action, when things like this don't work.

These were my exact steps as of 5 minutes ago:

1. Fresh clone from xplat.<br/>`git clone git@github.com:Microsoft/msbuild.git --branch xplat`
2. Verify the latest commit matches the branch head (c93ad0d05139c4b37b6448dfc89fc293d2cc5f65 as of this post).
3. Get into the clone's working directory, and run the build script.<br/>`./cibuild.sh --Target Compile`

The result is the exact same error. No `./Tools/corerun` file (See my comment from last week).

**Comment by radical** [comment]
And can you share the logs from this build?

**Comment by rainersigwald** [comment]
Ah, so the problem is that something transitively depends on an old DNX package:

```
EXEC : warning : Unable to find version '1.0.0-rc2-16177' of package 'dnx-coreclr-darwin-x64'. [/Volumes/DATA/Build/msb/buildme/build.proj]
  find: /Volumes/DATA/Build/msb/buildme/packages/dnx-coreclr-darwin-x64.1.0.0-rc2-16177: No such file or directory
```

And indeed that doesn't seem to be present on https://www.myget.org/feed/aspnetvnext/package/nuget/dnx-coreclr-darwin-x64

. . . but I tried deleting my `~/.nuget` folder, and the build passed without that package. And NuGet doesn't explain what the chain of dependencies was that led to it. So I'm still pretty confused.

**Comment by Sarabeth-Jaffe-Microsoft** [comment]
After performing the same steps @JunielKatarn followed, this is my output:
`Macintosh:msbuild appleprouduser$ ./cibuild.sh
Unsupported target detected: . Configuring as if for CoreCLR
Unsupported host detected: . Configuring as if for CoreCLR
Installing dotnet cli...
Restoring BuildTools version 1.0.25-prerelease-00231-07...
ERROR: Could not restore build tools correctly. See '/Users/appleprouduser/Dropbox/MSBuild/msbuild/init-tools.log' for more details.
Initializing BuildTools...
./init-tools.sh: line 84: /Users/appleprouduser/Dropbox/MSBuild/msbuild/packages/Microsoft.DotNet.BuildTools/1.0.25-prerelease-00231-07/lib/init-tools.sh: No such file or directory
Done initializing tools.
Build Command: ./Tools/corerun  ./Tools/MSBuild.exe ./build.proj /t:Build /p:OS=OSX /p:Configuration=Debug-NetCore /verbosity:minimal  /m /fl  "/flp:v=diag;logfile=./msbuild.log"
./cibuild.sh: line 42: ./Tools/corerun: No such file or directory`

And this is my init-tools.log file:
`Running: /Users/appleprouduser/Dropbox/MSBuild/msbuild/init-tools.sh
Installing 'https://dotnetcli.blob.core.windows.net/dotnet/beta/Binaries/1.0.0-beta-002173/dotnet-dev-osx-x64.1.0.0-beta-002173.tar.gz' to '/Users/appleprouduser/Dropbox/MSBuild/msbuild/Tools/dotnetcli/dotnet.tar'
Running: /Users/appleprouduser/Dropbox/MSBuild/msbuild/Tools/dotnetcli/dotnet restore "/Users/appleprouduser/Dropbox/MSBuild/msbuild/Tools/1.0.25-prerelease-00231-07/project.json" --packages /Users/appleprouduser/Dropbox/MSBuild/msbuild/packages --source https://dotnet.myget.org/F/dotnet-buildtools/api/v3/index.json
log  : Restoring packages for /Users/appleprouduser/Dropbox/MSBuild/msbuild/Tools/1.0.25-prerelease-00231-07/project.json...
error: Failed to retrieve information from remote source '/Users/appleprouduser/Dropbox/MSBuild/msbuild/packages'.
error: Unable to load the service index for source https://dotnet.myget.org/F/dotnet-buildtools/api/v3/index.json.
error:   The type initializer for 'Crypto' threw an exception.
error:   Unable to load DLL 'System.Security.Cryptography.Native': The specified module could not be found.
error:    (Exception from HRESULT: 0x8007007E)
Running: /Users/appleprouduser/Dropbox/MSBuild/msbuild/packages/Microsoft.DotNet.BuildTools/1.0.25-prerelease-00231-07/lib/init-tools.sh /Users/appleprouduser/Dropbox/MSBuild/msbuild /Users/appleprouduser/Dropbox/MSBuild/msbuild/Tools/dotnetcli/dotnet /Users/appleprouduser/Dropbox/MSBuild/msbuild/Tools
`

**Comment by rainersigwald** [comment]
@Sarabeth-Jaffe-Microsoft that looks like the original error @JunielKatarn had. Do _you_ see `./Tools/dotnetcli/shared/Microsoft.NETCore.App/1.0.0-rc2-23931/System.Security.Cryptography.Native.dylib`?

I created a new user on my Mac and was able to clone + `./cibuild` with no trouble.

**Comment by JunielKatarn** [comment]
@rainersigwald I'm not with my Mac right now, but at least on Ubuntu, I do see `Tools/dotnetcli/shared/Microsoft.NETCore.App/1.0.0-rc2-23931/System.Security.Cryptography.Native.so`.
Please note, Ubuntu has exactly the same symptoms as OS X.

**Comment by Sarabeth-Jaffe-Microsoft** [comment]
@rainersigwald I do have that file. (Using Mac OSX 10.11.4 in case that matters.)

**Comment by cdmihai** [comment]
I found this in the CLI issues: https://github.com/dotnet/cli/issues/1161

Do you have openssl? We might have found an implicit system dependency :)

**Comment by rainersigwald** [comment]
@cdmihai That does look awfully suspicious. `brew list --versions` tells me I have `openssl 1.02e_1` installed--and that would be machine wide, explaining my non-repro with a new user.

**Comment by JunielKatarn** [comment]
I have not installed brew on my Mac.
My Ubuntu installation is also pretty out-of-the-box, so it might also lack it.

Should a third-party software like brew be required at all to get MSBuild to compile?

**Comment by rainersigwald** [comment]
@JunielKatarn MSBuild doesn't (or at least shouldn't) have any dependencies on homebrew. However, since we use the `dotnet` CLI tool to download the packages we need (for compilers and build-and runtime references), we're at the mercy of any dependencies they choose to take.

In this case, the dependency is on having a dynamic OpenSSL library available; it shouldn't care about how that came about. `brew install openssl` and `apt-get install openssl` are easy, but you could also try [building from source](https://wiki.openssl.org/index.php/Compilation_and_Installation#Mac).

**Comment by Sarabeth-Jaffe-Microsoft** [comment]
Downloaded brew & updated openssl:

```
Macintosh:msbuild appleprouduser$ openssl version -a
OpenSSL 1.0.2g  1 Mar 2016
built on: reproducible build, date unspecified
```

**Comment by Sarabeth-Jaffe-Microsoft** [comment]
@rainersigwald suggested:

```
git clean -xdf
```

And it looks like it runs with no errors! Not sure if that command is all that's needed or if updating Openssl is necessary.

**Comment by JunielKatarn** [comment]
> apt-get install openssl are easy, but you could also try building from source.

@rainersigwald Ubuntu comes with openssl installed by default.
Actually, I have the exact same version as @Sarabeth-Jaffe-Microsoft in OS X (1.0.2g).

I'll try it on OS X later tonight.

**Comment by JunielKatarn** [comment]
I have confirmed that upgrading OpenSSL in OS X via homebrew solved the issue.
I would keep this issue upen until the workaround steps are added to the README file, or the transitive dependency goes away.

In short (@Sarabeth-Jaffe-Microsoft , please confirm):
1. Install homebrew from http://brew.sh.
2. `brew install openssl`
3. `brew link openssl --force` (to put homebrew's OpenSSL in front of Apple's OpenSSL).

The issue on Ubuntu is still not resolved.

**Comment by rainersigwald** [comment]
@Sarabeth-Jaffe-Microsoft `git clean -xdf` only cleaned up files that were downloaded/created in your repo. It worked around a problem (#590) that the initialization process decided (erroneously) that it was already complete, then failed later.

**Comment by rainersigwald** [comment]
@JunielKatarn 
> I would keep this issue upen until the workaround steps are added to the README file, or the transitive dependency goes away.

Agreed. Since we're probably stuck with the dependency, a clear error on our part might be nice too.

> `brew link openssl --force` (to put homebrew's OpenSSL in front of Apple's OpenSSL).

I don't think this is required (on my working Mac, `openssl version` returns `OpenSSL 0.9.8zh 14 Jan 2016` which is the system one).

> The issue on Ubuntu is still not resolved.
Ok, let's dig in further. On my working Ubuntu 14.04 system, I did this 
```diff
diff --git a/init-tools.sh b/init-tools.sh
index 033a652..7fa299d 100755
--- a/init-tools.sh
+++ b/init-tools.sh
@@ -75,7 +75,7 @@ if [ ! -e $__PROJECT_JSON_FILE ]; then
     if [ ! -e $__BUILD_TOOLS_PATH ]; then
         echo "Restoring BuildTools version $__BUILD_TOOLS_PACKAGE_VERSION..."
         echo "Running: $__DOTNET_CMD restore \"$__PROJECT_JSON_FILE\" --packages $__PACKAGES_DIR --source $__BUILDTOOLS_SOURCE" >> $__init_tools_log
-        $__DOTNET_CMD restore "$__PROJECT_JSON_FILE" --packages $__PACKAGES_DIR --source $__BUILDTOOLS_SOURCE >> $__init_tools_log
+        strace -f -e trace=file $__DOTNET_CMD restore "$__PROJECT_JSON_FILE" --packages $__PACKAGES_DIR --source $__BUILDTOOLS_SOURCE >> $__init_tools_log 2>&1
         if [ ! -e "$__BUILD_TOOLS_PATH/init-tools.sh" ]; then echo "ERROR: Could not restore build tools correctly. See '$__init_tools_log' for more details."; fi
     fi
```

And I see this around loading crypto:
```
[pid 13479] open("/home/vagrant/src/msbuild/Tools/dotnetcli/shared/Microsoft.NETCore.App/1.0.0-rc2-23931/System.Security.Cryptography.Native.so", O_RDONLY|O_CLOEXEC) = 120        
[pid 13479] open("/etc/ld.so.cache", O_RDONLY|O_CLOEXEC) = 120                                                                                                                     
[pid 13479] access("/etc/ld.so.nohwcap", F_OK) = -1 ENOENT (No such file or directory)                                                                                             
[pid 13479] open("/lib/x86_64-linux-gnu/libcrypto.so.1.0.0", O_RDONLY|O_CLOEXEC) = 120                                                                                             
[pid 13479] access("/etc/ld.so.nohwcap", F_OK) = -1 ENOENT (No such file or directory)                                                                                             
[pid 13479] open("/lib/x86_64-linux-gnu/libssl.so.1.0.0", O_RDONLY|O_CLOEXEC) = 120                                                                                                
[pid 13479] open("/dev/urandom", O_RDONLY|O_NOCTTY|O_NONBLOCK) = 120                                                                                                               
[pid 13479] open("/usr/lib/ssl/openssl.cnf", O_RDONLY) = 120
```

Can you make a similar change and post `init-tools.log`? Hopefully there's a glaringly obvious difference . . .

**Comment by rainersigwald** [comment]
On a clean Ubuntu 14.04 VM, I get this failure:
```
$ /home/vagrant/msbuild/Tools/dotnetcli/dotnet restore "/home/vagrant/msbuild/Tools/1.0.25-prerelease-00231-07/project.json" --packages /home/vagrant/msbuild/packages --source https://dotnet.myget.org/F/dotnet-buildtools/api/v3/index.json
Failed to load /home/vagrant/msbuild/Tools/dotnetcli/shared/Microsoft.NETCore.App/1.0.0-rc2-23931/libcoreclr.so, error: libunwind.so.8: cannot open shared object file: No such file or directory
Failed to bind to coreclr
```

This appears to have been resolved with `sudo apt-get install libunwind8`, so I think we have another hidden dependency (and needed update to the docs).

**Comment by Sarabeth-Jaffe-Microsoft** [comment]
Until the hidden dependencies are resolved, I updated the build instructions in the [README](https://github.com/Microsoft/msbuild/wiki/Building-Testing-and-Debugging-on-.Net-Core-MSBuild). Feel free to update it if I left anything out!

**Comment by cdmihai** [comment]
@Sarabeth-Jaffe-Microsoft 

Is libunwind8 for both mac and linux, or only linux?

**Comment by rainersigwald** [comment]
@cdmihai Looks like it's Linux and FreeBSD only, though I'm not sure how stack walking works on the Mac:

https://github.com/dotnet/coreclr/blob/775003a4c72f0acc37eab84628fcef541533ba4e/Documentation/building/osx-instructions.md mentions OpenSSL but not libunwind

https://github.com/dotnet/coreclr/blob/775003a4c72f0acc37eab84628fcef541533ba4e/Documentation/building/linux-instructions.md mentions `libunwind8`.

**Comment by Sarabeth-Jaffe-Microsoft** [comment]
@cdmihai @rainersigwald Got it, just clarified it in the documentation.

**Comment by JunielKatarn** [comment]
I still see the issue under Ubuntu. Can you reopen?

**Comment by rainersigwald** [comment]
@JunielKatarn What issue, exactly? Did you install `libunwind8`?

**Comment by JunielKatarn** [comment]
@rainersigwald 
Yes, I have `libunwind8` installed. Still getting the same error.
That said, I AM using Ubuntu 16.04, which may not be supported.
I'll try again with 14.04 and post the results here.

**Comment by cdmihai** [comment]
@JunielKatarn, can you please paste in the error you're getting? Hard to follow the thread to figure out which error you're referring to.

**Comment by rainersigwald** [comment]
@JunielKatarn I'm also interested in the error, but I bet it will be that you're using 16.04. AFAIK, .NET Core is only currently supported on 14.04 (though it [looks like](https://github.com/dotnet/coreclr/pull/4599) folks might be working on modernizing.

I added that explicitly to the xplat build instructions: https://github.com/Microsoft/msbuild/wiki/Building-Testing-and-Debugging-on-.Net-Core-MSBuild/7884f4d239eedb6a47e170530d031c7d7ab28313

**Comment by JunielKatarn** [comment]
@rainersigwald, @cdmihai 
I just set up a brand new Ubuntu 14.04 instance, installed `libunwind8`, and MSBuild got built successfully off the shelf.

I think it's safe to assume the issue lies with 16.04/corerun incompatibility.
